### ISSUE TITLE
chain overloading

### ISSUE THREAD

**Issue by jsmith** [supermacro/neverthrow#36]
Hey supermacro, thanks so much for this library :) I'm starting to use it in one of my projects and love it so far! Anyway I was just wondering if you considered using function overloading for the `chain` API? I'm imaging something similar to the `pipe` API in [`fp-ts`](https://github.com/gcanti/fp-ts/blob/master/src/pipeable.ts#L111).

Here is kinda what I'm thinking although this has not been tested:
```typescript
export function chain<T1, T2, E>(r1: Promise<Result<T1, E>>, r2: (v: T1) => Promise<Result<T2, E>>): Promise<Result<T2, E>>;
export function chain<T1, T2, T3, E>(r1: Promise<Result<T1, E>>, r2: (v: T1) => Ok<T2, E> | Err<T2, E> | Promise<Result<T2, E>>, r3: (v: T2) => Promise<Result<T3, E>>): Promise<Result<T3, E>>;
export function chain<T1, T2, T3, T4, E>(r1: Promise<Result<T1, E>>, r2: (v: T1) => Ok<T2, E> | Err<T2, E> | Promise<Result<T2, E>>, r3: (v: T2) => Ok<T3, E> | Err<T3, E> | Promise<Result<T3, E>>, r4: (v: T3) => Promise<Result<T4, E>>): Promise<Result<T4, E>>;
export function chain<T1, T2, T3, T4, T5, E>(r1: Promise<Result<T1, E>>, r2: (v: T1) => Ok<T2, E> | Err<T2, E> | Promise<Result<T2, E>>, r3: (v: T2) => Ok<T3, E> | Err<T3, E> | Promise<Result<T3, E>>, r4: (v: T3) => Ok<T4, E> | Err<T4, E> | Promise<Result<T4, E>>, r5: (v: T4) => Promise<Result<T5, E>>): Promise<Result<T5, E>>;
export function chain<T1, T2, T3, T4, T5, T6, E>(r1: Promise<Result<T1, E>>, r2: (v: T1) => Ok<T2, E> | Err<T2, E> | Promise<Result<T2, E>>, r3: (v: T2) => Ok<T3, E> | Err<T3, E> | Promise<Result<T3, E>>, r4: (v: T3) => Ok<T4, E> | Err<T4, E> | Promise<Result<T4, E>>, r5: (v: T4) => Ok<T5, E> | Err<T5, E> | Promise<Result<T5, E>>, r6: (v: T5) => Promise<Result<T6, E>>): Promise<Result<T6, E>>;
export function chain<T1, T2, T3, T4, T5, T6, T7, E>(r1: Promise<Result<T1, E>>, r2: (v: T1) => Ok<T2, E> | Err<T2, E> | Promise<Result<T2, E>>, r3: (v: T2) => Ok<T3, E> | Err<T3, E> | Promise<Result<T3, E>>, r4: (v: T3) => Ok<T4, E> | Err<T4, E> | Promise<Result<T4, E>>, r5: (v: T4) => Ok<T5, E> | Err<T5, E> | Promise<Result<T5, E>>, r6: (v: T5) => Ok<T6, E> | Err<T6, E> | Promise<Result<T6, E>>, r7: (v: T6) => Promise<Result<T7, E>>): Promise<Result<T7, E>>;
export function chain<T1, T2, T3, T4, T5, T6, T7, T8, E>(r1: Promise<Result<T1, E>>, r2: (v: T1) => Ok<T2, E> | Err<T2, E> | Promise<Result<T2, E>>, r3: (v: T2) => Ok<T3, E> | Err<T3, E> | Promise<Result<T3, E>>, r4: (v: T3) => Ok<T4, E> | Err<T4, E> | Promise<Result<T4, E>>, r5: (v: T4) => Ok<T5, E> | Err<T5, E> | Promise<Result<T5, E>>, r6: (v: T5) => Ok<T6, E> | Err<T6, E> | Promise<Result<T6, E>>, r7: (v: T6) => Ok<T7, E> | Err<T7, E> | Promise<Result<T7, E>>, r8: (v: T7) => Promise<Result<T8, E>>): Promise<Result<T8, E>>;

export async function chain(
  start: Promise<Result<any, any>>,
  ...transforms: any
): Promise<Result<any, any>> {
  return await (transforms as Array<(v: any) => Promise<Result<any, any>>>).reduce(
    async (current, f) => {
      const mapped = await (await current).asyncMap(f);
      return mapped.andThen((inner) => inner);
    },
    start,
  );
}
```

Let me know your thoughts!

**Comment by supermacro** [comment]
Hey @jsmith, 

Happy to hear that you're enjoying this lib!

It's funny you bring up this issue because I'm collaborating currently with a fellow github user on implementing a "thenable" `Result`. This new API would supersede the `chain` functionality.

Have a read through [this](https://github.com/supermacro/neverthrow/issues/32) issue, and take a peak at the [corresponding pr](https://github.com/supermacro/neverthrow/pull/34) which **should** get merged tomorrow or thursday and released to npm no later than friday (but as with anything in software, there may be delays!).

Does this new API solve your issue? I would appreciate any feedback.

**Comment by jsmith** [comment]
Yes, that will be perfect! I'm using `neverthrow` to define a set of express routes which use a combination of synchronous and asynchronous operations so this new API should be a good replacement. 

I'm building a remote git api and currently have something like this:
```typescript
router.post("/touch", async (req) => {
  return chain(
    getOrCloneRepo(env.git),
    getFullPath(req.body.gitPath),
    isInGitRepo,
    doesNotExist,
    parentExists,
    async ({ fullPath }) => {
      await fs.promises.writeFile(fullPath, "");
      return success();
    },
  ).then(unwrap);
});
```

But will just be able to use `ResultAsync` and `map`!

Anyway, thanks for your help :)

**Comment by supermacro** [comment]
@jsmith https://github.com/supermacro/neverthrow/releases/tag/2.3.0

Try it out. It's published to npm.

**Comment by jsmith** [comment]
@supermacro awesome!

TLDR
I have a `ResultAsync` and I want to and `andThen` with an `async` function that returns `Ok` or `Err`.
```typescript
```


I just have a question about the API of `ResultAsync` that came up while I was refactoring. I'll describe my use case first! 

Basically, I have several functions that currently return `Promise<Result<T, E>>`. An example might look like this:
```typescript
export const doesNotExist = async (
  opts: { fullPath: string },
): Promise<Result<{ fullPath: string }, { status: 400, error: string }>> => {
  const result = await checkStatus(opts.fullPath); // result is "file" | "folder" | "enoent"
  // note that error() just wraps a string in a { status: 400, error: string } object
  return result === "enoent" ? ok(opts) : error(`"${opts.fullPath}" already exists.`);
};
```

This function just checks that a file does not exist. It returns an error if it does exist otherwise it just passes information along to the next handler. Using the example from above, my handler now looks like ths:
```typescript
router.post("/touch", async (req) => {
  return getOrCloneRepo(env.git)
    .andThen(getFullPath(req.body.gitPath))
    .andThen(isInGitRepo)
    .andThen(doesNotExist)
    .andThen(parentExists)
    .andThen(async ({ fullPath }) => {
      await fs.promises.writeFile(fullPath, "");
      return success();
    })
    .then(unwrap);
});
```

Let me know if this makes sense! If not, I can create a simpler test case. Anyway the issue is that, for example, `doesNotExist` returns a `Promise<Result<T, E>>` which isn't currently supported with the API. The current interface of `ResultAsync` looks like this:
```typescript
export declare class ResultAsync<T, E> {
...
    andThen: <U>(f: (t: T) => Ok<U, E> | Err<U, E> | ResultAsync<U, E>) => ResultAsync<U, E>;
...
}
```

Would it be possible to modify the API to accept a `Promise<Result<U, E>>`?

```typescript
export declare class ResultAsync<T, E> {
...
    andThen: <U>(f: (t: T) => Ok<U, E> | Err<U, E> | Promise<Result<U, E>> | ResultAsync<U, E>) => ResultAsync<U, E>;
...
}
```

This would be similar to the `chain` API which allows you to return `Ok<T2, E> | Err<T2, E> | Promise<Result<T2, E>`. Does this make sense? Very sorry if I'm missing something obvious and thanks for your help :)

**Comment by paduc** [comment]
Hello @jsmith 

The solution to your problem is to have `doesNotExist` return a `ResultAsync` rather than a `Promise<Result>`. You can do this by wrapping your `checkStatus` promise using `ResultAsync.fromPromise`.

```typescript
export const doesNotExist = (
  opts: { fullPath: string },
): ResultAsync<{ fullPath: string }, { status: 400, error: string }> => 
  ResultAsync.fromPromise(checkStatus(opts.fullPath))
             .andThen(result => result === "enoent" ? ok(opts) : error(`"${opts.fullPath}" already exists.`))
```

Note that you might want to catch a possible error coming from `checkStatus`. In your example, you would have surrounded the call with `try/catch`. In my example, you can provide a second argument to `fromPromise` to handle the error.

**Comment by jsmith** [comment]
Sorry about that, I should have read through the documentation more carefully. I see now the power of `fromPromise`! Refactoring was going well but I encountered a type issue and I'm not entirely sure how to proceed. I'm going to keep looking at it but thought I would post it here to see if you had any thoughts!

Basically, ignoring the type errors from the first two lines of code, this type checks:
```typescript
import { Result } from "neverthrow" 
// @ts-ignore
const err1: Result<{}, { a: "a" }> = {};
// @ts-ignore
let err2: Result<{}, { a: "a" } | { b: "b" }> = {};
err2 = err1;
```

But using the same code with `ResultAsync` won't type check:
```typescript
// I've reduced ResultAsync to just have `andThen`
// It seems like Ok<U, E> and Err<U, E> are causing the issues
// But I haven't totally wrapped around why this is happening
export declare class ResultAsync<T, E> {
  // original is `andThen: <U>(f: (t: T) => Ok<U, E> | Err<U, E> | ResultAsync<U, E>) => ResultAsync<U, E>;`
  andThen: <U>(f: (t: T) => Ok<U, E>) => ResultAsync<U, E>;
}

// @ts-ignore
const err1: ResultAsync<{}, { a: "a" }> = {};
// @ts-ignore
let err2: ResultAsync<{}, { a: "a" } | { b: "b" }> = {};

// This next line fails with:
// Type 'TestResult<{}, { a: "a"; }>' is not assignable to type 'TestResult<{}, { a: "a"; } | { b: "b"; }>'.
//   Type '{ a: "a"; } | { b: "b"; }' is not assignable to type '{ a: "a"; }'.
//     Property 'a' is missing in type '{ b: "b"; }' but required in type '{ a: "a"; }'.ts(2322)
err2 = err1;
```

Pulling your example from #30 but replacing `Result` with `ResultAsync`, `ok` with `okAsync` and `err` with `errAsync`:
```typescript
import { ResultAsync, okAsync, errAsync } from "neverthrow";

type E1 = number;
type E2 = string;
interface E3 {
  errorCode: number;
  message: string;
}

type MyUnion = { type: "e1"; val: E1 } | { type: "e2"; val: E2 } | { type: "e3"; val: E3 };

type F1 = (sth: unknown) => ResultAsync<string, MyUnion>;
type F2 = (sth2: string) => ResultAsync<string, MyUnion>;
type F3 = (sth3: string) => ResultAsync<string, MyUnion>;

const f1: F1 = (sth: unknown) => {
  console.log(sth);
  const x = Math.random() * 10;
  return x > 5 ? okAsync("yes-one") : errAsync({ type: "e1", val: 1234 });
};

const f2: F2 = (sth: string) => {
  console.log(sth);
  const x = Math.random() * 10;
  return x > 5 ? okAsync("yes-two") : errAsync({ type: "e2", val: "noooo :(" });
};

const f3: F3 = (sth: string) => {
  console.log(sth);
  const x = Math.random() * 10;
  return x > 5
    ? okAsync("yes-three")
    : errAsync({ type: "e3", val: { errorCode: 500, message: "internal server error" } });
};

const result: ResultAsync<string, MyUnion> = f1("maybe").andThen(f2).andThen(f3);
```

You'll notice that this has type issues that will end up looking like this:
```
Type 'MyUnion' is not assignable to type '{ type: "e1"; val: number; }'.
```

Thanks again for your help 😅

**Comment by paduc** [comment]
@jsmith I pasted your code (the last code block, not the declaration of ResultAsync) et I have no type error. It works as expected and even without the type annotation on `result`, the type is `ResultAsync<string, MyUnion>`.

Am I missing something ?

**Comment by jsmith** [comment]
@paduc thank you for testing my code samples and for creating a more succinct snippet! I quickly tested and found that I only have this error when `strictFunctionTypes` (or `strict`) is set to `true` in my tsconfig. This makes sense as when `strictFunctionTypes` is set to `false`,  it "disables bivariant parameter checking for function types.". Maybe you don't have this option turned on?

**Comment by paduc** [comment]
@jsmith Indeed I had neither `strict` nor `strictFunctionTypes` set to `true`. I wasn't even familiar with their existence until today (thanks!). 

I turned it on but I didn't get a type error... 

I copied the lib and your examples in a typescript playground, no warning/errors: [try it here](https://www.typescriptlang.org/play/#code/C4TwDgpgBAShDOBXANsAPAFQDRQKID4oBeKAeQGtMcCoAfPAJwar3wCg2B6TqBZASwB2wALQATfvACGAI2QQRgiAA9RApVAACoSPADGDfmFF8hwToID2IxPAUyIAM0sMFYp0Ihs9lwfGBQluTEUCwEABQAblLIiBAAXFAYAJSJFGGERIRKAO5k5FExcckc3Lzw6qIS0nIKSqoi6tDa4AgGRiYVZhbWtvZOLm4eSt6+-rxMIRnhEEyJuKmMzNisxNkQebhMM0wl3shS8PD5GVAA3mxQUD5+wAyIesAu4a5SYr7IIFDRsQlJyecAL4cK6SCjhRbAAAWkigsPSKxoFyuV1cwEQDEEUDucUuUGBeMkWwYEMS0NhsOJp2RKLRGKxAEJyfAAHRggolK4Eq4AWykYDQAEF8OFHIlwsBEik1lBBYs4EhUELqIQaaiIOjMYECo4JTDWT9ipz8SCoGVTMJxJJZPJFCo1J4tDo2oZjCILeYrDZBH0xCJogx4Hi+WAqQBVEUAfTFUBm8wBWSgYflCBQ6BWEfOePVmqxQT1khZhogxu5Zp4bzEcMEzkCWKhljyzLhx1eyE+UD6jhQUGcDGuvkis2AQgA5lApHiygq02gZ0rEQAmfDUACM7CuZTMlgnsFTC+oy7xUkEYgwUIggjQEdF4sl-xl8+AgvgIEEemvKpTiufr-fn9YY9T3PS9rxFGMJSlBNCCfACCG-WcwxVICzwvK8bwg+9pUTWCkNWegnxfN8Pzw+DElwlU6D3H8iP-UjVWzKA6S1XVmSLIoSzxMtDmIgBZfkwNvWMsOgqAAAUGEsHlJAgMCENQWiSMotUmI1elqLTRSWUcSSeQkqSZNFAsDQ45JS1Nc0uktaobTqe1GkdFpdHaN0PR6b1fX9KRA2DKRgD0KEpmFcIgjvKCZUFHBI1mBhxT+BYIsWQUZRU5i8wKNji3MvFI0QPwpEcCAw0EHIGH5UkkizFFVNzbF9XY34uNNXL8sK4rSv5YkKtwKqUWhSS8lyJZngAcgAYRieQqwAAxa6Q2pKsrQyYaa6wnLEKBG8yCT0A4jiWak8Rufx7keZ5XneQQOxilx4yBU12Qq5t4UoREGOqtLexiOwmsJeAushfUWwOt7epzdSmXqx7ss3HgPStGpbXqB0NCcl0OndKzPV6H07D9AMg15ATgujMKH0TOVyP3dBItWVK1K1GLjJZG6GBhqAQ3DcC4ruxNkypn8WEzenaqZ1j6tZszfquE9UNAm9SeE8KcOpuD8Hk9B6KnHgAGUhD0aBo2B-hRysVwqz7JYcGhaBPudYGYhyKQQGOKQNNQFCQPQqNMOVmDqcUtWNbV7XykqBHbLtBomidVp9FdTpKncvLPIJz20MExXIPJ-3BfoqjCL-JTWGD-ORfUpm2Ml6XyzDswI9qKOUeaZ144xtyvRTvGvJ8mWi-4gUFd9nPxN0mS5IFzSi7VsGaorphA-o5nq65U0+X8wKSCFKNQqVkfaZinm8FEynZRSxjPsriWmBcdm5oKorFvKxYMFn-rGygIbiVGib2wgGb74LQ6mAVavh1pLC2jXQBj9gEA3mLPO219JIMCajtPaxxC7EUOlcMAhhojAENrggydhEj6WknYOcqs3rsCOmMHEZ0SSuHgKQseFDYLUIBCpNikYiHkOgCQJhNd-B+X4HoXsukyHjzeuEXhMkWHENkhgFcExkEADFBAAH4j55XIFYHIghRILEngpaeoMVLyACLkSRJDR4KMoYLahIRZF2BZDbQQ4RCi-D9p-DY+RPFGkYvwRwsZWbqM4YxK4VjWH8J8TkaxEAWR6D8gFGYMov5MGmKE9xJYpbVWBB9BmWIhqYPfOEKJCj2YhiCtzPe2FCDJXoPE7e0FGLFIDkXcIES6qFh4dE1xaFwg8XfLGJhxj0xfnPtVaqQSRkIDZP9bYZlZ5TLnlqdJzBaYRCYSzG+bMumr32asopvj0ibJFFIJ2-AAi6m2VlY0eTkhYEYnsImy1+xbwwkfBKfMqJNIjC06qbSaIdK6dw5xCS3GDKLrM5h7txkl0mSsuEwSXhzOhuEpFtJCmxJOBmEUtzTKHLLCsz66yhbnMudc1FrJJb3JRICR5zzTSyy9lMT5tTRIUXwnCxeEysitN8SUvQnSVlgr6ZCphiKVkzOpfMgGGLMVHJxeGFUsqV5IuJVM46liNgADUOIhBuXMrKpoSXYtyPq341ZhHvggJYYJQqoAaNiZauILJekKKgIkC1HEukMqedVF5HNkmb1CMFXe2c6myiiofEJvN6miTFQo-pl5qUym2evFJQQY27CDWUfi5AEA8uhRcqQVzbJ4jcdU8ISA9AGyOL-ZAMgpB6HIOKUZcKMgn0TfVD1fCU3uNrfW+Ajbm2tpKDtOh2pFJTGkcWKCYzA6g35ZEwV7TiLhHiSyJhlhkBDjKSc16qq7l7G1SomdW9pGxqMcWrBy6Nw4qFZuvpO690QAPZsDJV7c17FKJwLgPAACalgMS8GUFIHkYB5DHGEQwAIF5XAAY4KMW4KjVyLunmcQEOAzgTkSCNKQI18SZHPR03DUh8OEfxCUM9MVFwYawVhnDeGoAEaI4CKiuGZD4ZkOxkjMVVwcHtrgVcIQfQ8gcCg4Ti4QgnTHGwMwsxHAtugLgAAzL1VmY1LDuG9YgCTsxgwIGkKOP4cnBCjjYASe2vEQDFX4GAkguHnT4YgKuEaABub4MR5iiY4-QZzrRXOLk8955A8wZP+fONiILrGIBqdCz8eYGnrOtCgKo0TJAa3QkSDovRBjHzrv-OZ0cOBbP2d8Owe2qiZNZf8FCejUASucqKx+ErZW7OCAc4IKraXVEabq9CNTiRmuFeBVg9rUBytdcqxwM9jh0Ppcy7GeruXBC6MbAVxMNJjq7oScgSwo5stQmNGe5QIR+LQm3bLKSEIoAACooCrgAAx4k+udwgABWJ106OkjRAAgEQvgIBbS9aRjdgXICufczgJLT3FxqYACzUas3NqdjhGs1ZCMdkbdwxyiR22MPbLIDtHfq6dqd52SCXahNd08t2ASPZe297FH2oDfedUERS4R-uA+ADkSwoPEgxW55Dv4I0IAhdhz51jVhLDy69TzlHk7UOOGG+lgbK2ctNbxxZgntC-DE9J8dinqGqdTb8rTsq9OeR3aZ698GWo2efcYpz8g3PefwBEP1CAIO6XC4XmRmLUO4sJel+F6LWmdN-E+8957OAeTGakKZ-DinMQxCa7MIc-ZWbseV2j1DTC0wMeK7r0rU3OvdZIwtnnfIQAOEgVcFkLKBkY+NM34Cre1MAhREAA)

**Comment by jsmith** [comment]
No problem :) They are very useful TS options! I typically have `strict` enabled for all my projects. I'm not sure what settings there are in the playground but here is a [codesandbox](https://codesandbox.io/s/heuristic-hoover-v9kxc?file=/src/index.ts) with the code from the playground! I should have sent this earlier, sorry about that!

**Comment by supermacro** [comment]
@jsmith @paduc I'm currently at work so I can't really spend the time thinking about this, but here's a new github issue: https://github.com/supermacro/neverthrow/issues/39

Let's move the convo there.
